### ISSUE TITLE
Conjugation of numbers not working

### ISSUE THREAD

**Issue by jebej** [symengine/SymEngine.jl#110]
It seems that conjugation does not work, as it fall back on the `conj(x) = x` Julia base definition:
```
julia> @vars a; conj(1im*a)
I*a
```

**Comment by isuruf** [comment]
you can define `conj(x::Basic) = real(x) - 1im * imag(x)` for now

**Comment by jebej** [comment]
Is this behavior meant to be handled in the C++ code?  Is the `Basic ` number type meant to be a real number?

**Comment by isuruf** [comment]
> Is the Basic number type meant to be a real number?

No, it's not. It's just that `conj` is not defined for `Basic` yet. (Because nobody needed `conj` until now)

**Comment by jebej** [comment]
Ok, but then your definition won't work, since `real(1im*a) = I*a` currently. It also seems that `imag(1im*a)` currently throws an error.

**Comment by isuruf** [comment]
Ah, `real` and `imag` are not working properly. This needs to be fixed. One problem is that we don't know whether `a` is real or not.

**Comment by mikolajpp** [comment]
Any progress on this? 
I would very much like to have it. Currently doing algebra over complex numbers 
is incovenient without this working on more complex types like matrices.

**Comment by timkittel** [comment]
@isuruf is there a plan for complex number operations to be implemented? I would really like to switch to `SymEngine`.

**Comment by isuruf** [comment]
There are no plans yet. Let's figure out what needs to be implemented.

Is it only a `Real` and `Imag` class so that `conj(a)` gives `Real(a) + I*Imag(a)` as output ?

**Comment by GiggleLiu** [comment]
> There are no plans yet. Let's figure out what needs to be implemented.
> 
> Is it only a `Real` and `Imag` class so that `conj(a)` gives `Real(a) + I*Imag(a)` as output ?

We need these two functions desperately in [YaoSym.jl](https://github.com/QuantumBFS/YaoSym.jl/blob/fix-conj-and-expect/src/symengine/blocks.jl), otherwise, I need to figure out how to compute the imaginary part of a general `sin(x)`, which is possible, but with a pretty long expression.

**Comment by isuruf** [comment]
@GiggleLiu, can you describe the functionalities you need from a `Real` and `Imag` class?

**Comment by GiggleLiu** [comment]
It is simple, just “real(sin(x)) -> Re(sin(x))”. Just a function.

**Comment by isuruf** [comment]
Does https://github.com/symengine/SymEngine.jl/blob/master/test/runtests.jl#L212 work for you?

**Comment by GiggleLiu** [comment]
@isuruf Exactly, how can I miss it 👍 . Still, I’d wish it can become a standard function in the code base of SymEngine, and deserves a special BasicType for dispatch. (Btw: possible to assign a BasicType?)

**Comment by GiggleLiu** [comment]
I tried the suggested `SymFunction`, it works fine. Do you have any suggestion to define the evaluation function when the number is known, e.g. `Re(0.5)` to `0.5`.
